### PR TITLE
Add RPCA and standalone RCLR functionality

### PR DESCRIPTION
**Pull request by cameronmartino**
This PR merges the repeated functionality of DEICODE which included the rclr transformation and RPCA functions. All previous testing used for RPCA/rclr in DEICODE has been adapted here. The tutorials for DEICODE have been added and altered to function through gemelli. Additionally, the rclr transformation as a standalone command and QIIME2 command has been added (a previously requested functionality from DEICODE). Overall, this prevents any shared functionality from being repeated across repos and causing issues later. This also allows gemelli to function on both cross-sectional and repeated measure experimental setups. The tutorials have been set up this way to minimize confusion on when to use CTF or RPCA. 

### Features

* Robust Aitchison PCA
    * with manually chosen n_components
    * with auto chosen n_components
    * associated tests added for function, standalone, & QIIME2
* Robust centered log-ratio (rclr) transformation
    * allows users to use the transform alone outside CTF/RPCA
    * associated tests added and extra tests for rclr added  
* Tutorials
    * Cross-sectional for RPCA (standalone & QIIME2)
    * Repeated measure for CTF (standalone & QIIME2)

### Miscellaneous

* auto_rpca rank estimation min rank now 3 
* added citations for CTF & RPCA (linked when appropriate)
* updated the function descriptions

### CODE REVIEW

File: gemelli/_rpca_defaults.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+# Configuration file where you can set the parameter default values and
+# descriptions. This is used by both the standalone RPCA and QIIME 2 RPCA sides
+# of gemelli.
+
+DEFAULT_RANK = 3
```
Comment (gwarmstrong) [inline gemelli/_rpca_defaults.py:13]: Are any of these shared with ctf? May make sense to have a single file with defaults if you envision wanting modify  a single default/description that is shared by ctf/rpca

File: gemelli/ctf.py [hunk 2]
```diff
@@ -3,8 +3,8 @@
 from pandas import concat
 from pandas import DataFrame
 from skbio import OrdinationResults, DistanceMatrix
-from gemelli.factorization import TensorFactorization
-from gemelli.preprocessing import build, rclr
+from gemelli.tensor_factorization import TensorFactorization
```
Comment (gwarmstrong) [inline gemelli/ctf.py:6]: `factorization` -> `tensor_factorization` and `rclr` -> `tensor_rclr` are kind of huge breaking API changes. Definitely requires a major version.

I would be worried if now `gemelli.factorization` now automatically refers to matrix factorization or something that. If anyone uses your API, the error message they see will be along the lines of "`TensorFactorization` does not exist in `gemelli.factorization`" as opposed to "`TensorFactorization` has moved from `gemelli.factorization` to `gemelli.tensor_factorization`"

File: gemelli/matrix_completion.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import numpy as np
+from gemelli.optspace import OptSpace
+from .base import _BaseImpute
+from scipy.spatial import distance
+
+
+class MatrixCompletion(_BaseImpute):
```
Comment (gwarmstrong) [inline gemelli/matrix_completion.py:15]: I am going to assume this is copy and pasted from `deicode`

File: gemelli/matrix_completion.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+        """
+
+        self.n_components = n_components
+        self.max_iterations = max_iterations
+        self.tol = tol
+
+        return
+
+    def fit(self, X):
```
Comment (gwarmstrong) [inline gemelli/matrix_completion.py:95]: Sklearn transformers `fit`, and `fit_transform` methods typically take an optional `y=None` argument (even if `y` is not used in the computation) [[see here](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html)]. Including this argument makes the logic around chaining together multiple models (e.g., fitting deicode, then using it's components in an ML) a lot easier to use, especially if using some of sklearn's super useful pipeline functionality.

File: gemelli/base.py [hunk 5]
```diff
@@ -23,6 +23,10 @@ def fit(self):
     def label(self):
         """ Placeholder for fit this
         should be implemetned by sub-method"""
+    def transform(self):
```
Comment (gwarmstrong) [inline gemelli/base.py:26]: I would be __super__ hesitant to simply return `sample_weights` as the `transform` method like this.

Generally the expected behavior of `transform` is "I give you data, and you transform it for me". Your `transform` method is not doing that. Furthermore, `sample_weights` is already accessible as an attribute and you are not doing any extra work on it.

Side note: Is there any way to project new (i.e., not used in  fitting procedure) samples into RPCA/CTF space?

File: gemelli/matrix_completion.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        # save the solution (of the imputation)
+        self.solution = self.U.dot(self.s).dot(self.V.T)
+        self.eigenvalues = np.diag(self.s)
+        self.explained_variance_ratio = list(
+            self.eigenvalues / self.eigenvalues.sum())
+        self.distance = distance.cdist(self.U, self.U)
+        self.feature_weights = self.V
+        self.sample_weights = self.U
+
+    def fit_transform(self, X):
```
Comment (gwarmstrong) [inline gemelli/matrix_completion.py:151]: Same as above

File: gemelli/preprocessing.py [hunk 7]
```diff
@@ -8,10 +8,11 @@
 
 import warnings
 import numpy as np
+from biom import Table
 from .base import _BaseConstruct
 
 
-def rclr(T):
+def tensor_rclr(T):
```
Comment (gwarmstrong) [inline gemelli/preprocessing.py:15]: Same as above, be careful with name changes.

File: gemelli/preprocessing.py [hunk 8]
```diff
@@ -94,12 +97,13 @@ def rclr(T):
...
-            M_rclr = rclr_matrix(M)
+            M_tensor_rclr = rclr(M)
+        M_tensor_rclr[~np.isfinite(M_tensor_rclr)] = 0.0
         # reshape to former tensor and return tensors
-        return M_rclr.reshape(T.shape).transpose(reverse_T)
+        return M_tensor_rclr.reshape(T.shape).transpose(reverse_T)
 
 
-def rclr_matrix(M):
+def rclr(M):
```
Comment (gwarmstrong) [inline gemelli/preprocessing.py:106]: Here it is. I would strongly caution avoiding this name change. `matrix` may make sense as the default. BUT switching the default will strongly affect any user code. Would probably suggest keeping `rclr_matrix`, and `rclr_tensor`, and keeping `rclr` as `tensor`. Possible to deprecate, but I would not just go around switching the API.

File: gemelli/q2/tests/simulations.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
+from __future__ import division
+# utils
+import pandas as pd
+import numpy as np
```
Comment (gwarmstrong) [inline gemelli/q2/tests/simulations.py:4]: Can I suggest a different location for this code than a `tests` directory? While used in your testing, I could see this code being pretty useful to others outside of a 'unit testing' context, in which it feels weird to import from `tests`.

Maybe something like `gemelli.simulations`?

File: gemelli/q2/tests/simulations.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+# blocks
+from scipy.stats import norm
+from numpy.random import poisson, lognormal
+from skbio.stats.composition import closure
+# Set random state
+rand = np.random.RandomState(42)
+
+
+def Homoscedastic(X_noise, intensity):
```
Comment (gwarmstrong) [inline gemelli/q2/tests/simulations.py:13]: suggested name change:
`Homoscedastic` -> `make_homoscedastic`, `Heteroscedastic` -> `make heteroscedastic`.
Kind of inline with `sklearn`'s `make_classification`, etc. https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_classification.html

Comment (gwarmstrong) [reply gemelli/q2/tests/simulations.py:13]: Cancel that. (after reading rest of code in this module)

File: gemelli/q2/tests/test_optspace.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+            np.testing.assert_array_almost_equal(
+                abs(V_exp[:, i]), abs(opt.feature_weights[:, i]))
+            np.testing.assert_array_almost_equal(abs(U_exp[:, i]),
+                                                 abs(U_res[:, i]))
+            np.testing.assert_array_almost_equal(abs(s_exp[:, i]),
+                                                 abs(s_res[:, i]))
+            np.testing.assert_array_almost_equal(abs(V_exp[:, i]),
+                                                 abs(V_res[:, i]))
+
+    def test_OptSpace_rank_raises(self):
```
Comment (gwarmstrong) [inline gemelli/q2/tests/test_optspace.py:73]: could ease debugging if you split these into different tests

File: gemelli/q2/tests/test_rpca_method.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+from qiime2.plugins import gemelli as q2gemelli
+from gemelli.rpca import rpca, auto_rpca
+from gemelli.scripts.__init__ import cli as sdc
+from gemelli.q2.tests.simulations import build_block_model
+from click.testing import CliRunner
+from nose.tools import nottest
+
+
+@nottest
+def create_test_table():
```
Comment (gwarmstrong) [inline gemelli/q2/tests/test_rpca_method.py:19]: This looks like duplicated code

File: gemelli/scripts/__init__.py [hunk 13]
```diff
@@ -1,3 +1,22 @@
...
+
+
+@click.group()
+@click.version_option(version=__version__)
+@click.pass_context
+def cli(ctx):
+    ctx.call_on_close(_terribly_handle_brokenpipeerror)
+
+
+import_module('gemelli.scripts._standalone_transforms')
```
Comment (gwarmstrong) [inline gemelli/scripts/__init__.py:20]: I am not sure I understand these imports

Comment (gwarmstrong) [reply gemelli/scripts/__init__.py:20]: [This link](https://stackoverflow.com/questions/34643620/how-can-i-split-my-click-commands-each-with-a-set-of-sub-commands-into-multipl) may be useful for avoiding the circular import anti-pattern

File: gemelli/scripts/tests/test_standalone_rpca.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+        # Check that the distance matrix matches our expectations
+        assert_array_almost_equal(dist_res.values, dist_exp.values)
+
+        # Check that the ordination results match our expectations -- checking
+        # each value for both features and samples
+        assert_ordinationresults_equal(ord_res, ord_exp)
+
+        # Lastly, check that gemelli's exit code was 0 (indicating success)
+        try:
```
Comment (gwarmstrong) [inline gemelli/scripts/tests/test_standalone_rpca.py:51]: This code block is repeated a lot. Possible to encapsulate it?

E.g.,

```
class CliTestCase(unittest.TestCase):

    def assertExitCode(self, value, result):
        try:
            self.assertEqual(0, result.exit_code)
        except:
            ...
```

And then inheriting `CliTestCase` in all of your CLI tests, and calling `assertExitCode`

File: setup.py [hunk 15]
```diff
@@ -84,7 +84,7 @@ def run(self):
     hit = _version_re.search(f.read().decode('utf-8')).group(1)
     version = str(ast.literal_eval(hit))
 
-standalone = ['gemelli=gemelli.scripts._standalone_ctf:standalone_ctf']
+standalone = ['gemelli=gemelli.scripts.__init__:cli']
```
Comment (gwarmstrong) [inline setup.py:87]:
```suggestion
standalone = ['gemelli=gemelli.scripts:cli']
```

File: gemelli/tensor_factorization.py [hunk 16]
```diff
@@ -521,6 +524,16 @@ def tenals(tensor,
     tensor_dimensions = tensor.shape
     # Frobenius norm initial for ALS minimization.
     initial_tensor_frobenius_norm = norm(tensor)**2
+    # rank est. (for each slice)
```
Comment (gwarmstrong) [inline gemelli/tensor_factorization.py:527]: This doesn't seem super in scope, ffr

**Review by gwarmstrong — CHANGES_REQUESTED** [review]
I have some comments scattered throughout, mostly about structure and repeated code.

File: gemelli/matrix_completion.py (hunk 3 above)
Comment (cameronmartino) [reply gemelli/matrix_completion.py:15]: Correct.

File: gemelli/_rpca_defaults.py (hunk 1 above)
Comment (cameronmartino) [reply gemelli/_rpca_defaults.py]: These have been merged into one file and reduced.

File: gemelli/ctf.py (hunk 2 above)
Comment (cameronmartino) [reply gemelli/ctf.py]: Reverted the changes back to `factorization `

File: gemelli/matrix_completion.py (hunk 4 above)
Comment (cameronmartino) [reply gemelli/matrix_completion.py]: y=None is now added. Thanks!

File: gemelli/base.py (hunk 5 above)
Comment (cameronmartino) [reply gemelli/base.py]: These have been removed or altered to match the PCA function in sklearn.

File: gemelli/matrix_completion.py (hunk 6 above)
Comment (cameronmartino) [reply gemelli/matrix_completion.py]: ditto

File: gemelli/preprocessing.py (hunk 7 above)
Comment (cameronmartino) [reply gemelli/preprocessing.py:15]: these have been changed to `matrix_rclr` and `tensor_rclr`

File: gemelli/preprocessing.py (hunk 8 above)
Comment (cameronmartino) [reply gemelli/preprocessing.py]: these have been changed to matrix_rclr and tensor_rclr

File: gemelli/q2/tests/simulations.py (hunk 9 above)
Comment (cameronmartino) [reply gemelli/q2/tests/simulations.py]: The simulations have been moved to the main package.

File: gemelli/q2/tests/test_optspace.py (hunk 11 above)
Comment (cameronmartino) [reply gemelli/q2/tests/test_optspace.py]: these have been split. Thanks!

File: gemelli/q2/tests/test_rpca_method.py (hunk 12 above)
Comment (cameronmartino) [reply gemelli/q2/tests/test_rpca_method.py:19]: Now not duplicated. Thanks!

File: gemelli/scripts/tests/test_standalone_rpca.py (hunk 14 above)
Comment (cameronmartino) [reply gemelli/scripts/tests/test_standalone_rpca.py]: Good idea, this is now implemented. Thanks!

**Comment by gwarmstrong** [comment]
Do we need to adjust the `setup.py package_data` or `MANIFEST.in` in order to make sure the test data ships with the package?

**Pull request merged**